KEP-5732: Topology-aware workload scheduling

44past4 opened the pull request:
- One-line PR description: Initial version of Topology-aware workload scheduling KEP

- Issue link: #5732

/sig scheduling

44past4 committed:
Topology-aware workload scheduling KEP

44past4 committed:
Fixed Toc

44past4 committed:
Added KEP reviewers and approvers

wojtek-t commented on keps/sig-scheduling/5732-topology-aware-workload-scheduling/README.md:
```diff
@@ -0,0 +1,990 @@
+    // Level specifies the key of the node label representing the topology domain.
+    // All pods within the PodGroup must be colocated within the same domain instance.
+    // Examples: "topology.kubernetes.io/rack"
+    Level string
```
Let's specify what happens if a PodGroup is replicated (there are multiple PodGroup instances/replicas).
I'm assuming that for each of those domain at that level is chosen, but there is not coordination between those (i.e. different pod group instances may be scheduled in different domains, but they can also share them).

wojtek-t commented on keps/sig-scheduling/5732-topology-aware-workload-scheduling/README.md:
```diff
@@ -0,0 +1,990 @@
+type DRAConstraint struct {
+    // ResourceClaimName specifies the name of a specific ResourceClaim
+    // within the PodGroup's pods that this constraint applies to.
+    ResourceClaimName *string
```
What does ResourceClaimName mean if a given PodGroup is replicated (there are multiple podgroup instances/replicas)?

This would effectively mean sharing the same RC across multiple instances, which in many cases would be highly misleading.
However, arguably there can be usecases for it too, but then the algorithm effectively should consider all podgroup instances in a single round, but for that we don't know how many groups we even have.
@macsko - FYI (as this is slightly colliding with the kep-4671 update) 

So thinking about that more, I'm wondering if we can introduce that without further enhancing the API now (i.e. adding the `replicas` field to PodGroup).

Another alternative would be to very explicitly split the pod-group-replica constraints from the constraints across all pod-group-replicas and (at least for Alpha) focus only on the former.
So something more like (exact names and structures to be refined):
```
type PodGroupAllReplicasSchedulingConstraints {
  ResourceClaimName *string  // This one is supported only if Replicas=1
}

type PodGroupReplicaSchedulingConstraints {
  ResourceClaimTemplateName *string // Separate RC is created from this template for every replica.
}
```

wojtek-t commented on keps/sig-scheduling/5732-topology-aware-workload-scheduling/README.md:
```diff
@@ -0,0 +1,990 @@
+   with fallbacks (e.g., prefer Rack, fallback to Block). This would introduce
+   a Rank field to the Placement struct.
+
+2. **Optional/Preferred Scheduling Constraints:** Constraints that serve purely
```
I would make it out-of-scope for this KEP even for further releases - we can do that in a follow-up KEP if needed.

wojtek-t commented on keps/sig-scheduling/5732-topology-aware-workload-scheduling/README.md:
```diff
@@ -0,0 +1,990 @@
+2. **Optional/Preferred Scheduling Constraints:** Constraints that serve purely
+   as scoring mechanisms without hard requirements.
+
+3. **Multi-level Scheduling Constraints:** Handling nested constraints (e.g.,
```
I guess rephrasing these two comment - I think the extensions generally make sense to me, but I would claim that we should make them explicitly non-goals for this KEP and mark them more as future follow-up extensions.

44past4 commented on keps/sig-scheduling/5732-topology-aware-workload-scheduling/README.md:
```diff
@@ -0,0 +1,990 @@
+   Block -> Rack). This would involve iterative placement generation and a
+   Parent field in the Placement struct.
+
+4. **Pod Group Replicas Support:** Optimizing scheduling for identical
```
Yes, this is correct. Replicas will be supported but they will be scheduled one by one. With some changes to the algorithm we can try to optimize this process but this is out of scope for this KEP. I will rephrase it to make it clear.

nojnhuh commented on keps/sig-scheduling/5732-topology-aware-workload-scheduling/README.md:
```diff
@@ -0,0 +1,990 @@
+
+    // ResourceClaimTemplateName specifies the name of a ResourceClaimTemplate.
+    // This applies to all ResourceClaim instances generated from this template.
+    ResourceClaimTemplateName *string
```
How do these fields relate to the ResourceClaim references that Pods already have? What happens if the sets of claims referenced by a Workload and its Pods are different?

44past4 replied:
I have updated the KEP and removed the support for DRA based constraints from the alpha version and for beta I have proposed to wait for the KEP-5729: DRA: ResourceClaim Support for Workloads to define required API and lifecycle for the PodGroup level ResourceClaims.

wojtek-t commented on keps/sig-scheduling/5732-topology-aware-workload-scheduling/README.md:
```diff
@@ -0,0 +1,990 @@
+type PodGroupSchedulingConstraints struct {
+    // TopologyConstraints specifies desired topological placements for all pods
+    // within this PodGroup.
+    TopologyConstraints []TopologyConstraint
```
Does multiple topology constraints actually make sense here? What would be the usecase for it?

44past4 replied:
There are 2 main use cases for defining multiple topology constraints which I can see right now:

- when node label values are not unique among all nodes - for instance racks have indexes which are unique only within a given block - in this case we would like to be able to provide both of those labels as required constraints

- when some constraints are optional / best effort - this would require to introduce another field to TopologyConstraint which would allow specifying a given constraint as optional / best-effort.

wojtek-t replied:
The later I would actually expect "TopologyPreferences" field (or something like that) so I don't think that convinces me.

But the first usecase is interesting - I would actually mention it in the KEP explicitly.

Maybe we should actually explicitly mention in the API comment that in huge majority of cases we expect exactly 1 item in this list and mention this example as a potential exception.

sanposhiho replied:
> when node label values are not unique among all nodes - for instance racks have indexes which are unique only within a given block - in this case we would like to be able to provide both of those labels as required constraints

I don't understand why. In that case, they are supposed to just use `rack-label-existing-only-in-specific-blocks` on level, and then the scheduler should ignore the nodes on different blocks because those blocks don't have this label, no?

wojtek-t commented on keps/sig-scheduling/5732-topology-aware-workload-scheduling/README.md:
```diff
@@ -0,0 +1,990 @@
+// PodSetInfo holds information about a specific PodSet within a PodGroup,
+// primarily the list of Pods.
+// This struct is designed to be extensible with more fields in the future.
+type PodSetInfo struct {
```
Should PodSetInfo then have a link to PodGroupInfo?

I think we will need a way to iterate over podsets within podgroup and this seems like a good place to allow for it.

macsko commented on keps/sig-scheduling/5732-topology-aware-workload-scheduling/README.md:
```diff
@@ -0,0 +1,990 @@
+- **Action:** Iterate over distinct values of the topology label (TAS) or
+  available ResourceSlices (DRA).
+
+- **Output:** A list of Placement objects.
```
Will this phase generate all possible placements or only a subset of those? To reduce scheduling latency, in pod-by-pod scheduling we limit the number of feasible nodes to process (percentageOfNodesToScore setting). It might be important when the placements could be overlapping, what would cause otherwise some heavy processing.

44past4 replied:
This phase is assumed to generate all possible placements. However what we could do is to stop checking them in phase 2 if enough feasible placements has been found. The minimum number of feasible placements could be configurable and it could limit the scheduling latency in case of large number of placements. However I would treat this as a future optimization which could be done after some scalability testing.

johnbelamaric commented on keps/sig-scheduling/5732-topology-aware-workload-scheduling/README.md:
```diff
@@ -0,0 +1,990 @@
+
+    // DRAConstraints specifies constraints on how Dynamic Resources are allocated
+    // across the PodGroup.
+    DRAConstraints []DRAConstraint
```
I think that makes sense. Decoupling can help execution. We would treat the lifecycle and allocation of RCs in #5729. Allocation implies the constraint. #5194 should also merge with #5729, I think. It was conceived prior to the existence of the Workload API and I think #5729 encompasses a more holistic set of functionality.

44past4 replied:
I have updated the KEP and removed the support for DRA based constraints from the alpha version and for beta I have proposed to wait for the KEP-5729: DRA: ResourceClaim Support for Workloads to define required API and lifecycle for the PodGroup level ResourceClaims.

erictune commented on keps/sig-scheduling/5732-topology-aware-workload-scheduling/README.md:
```diff
@@ -0,0 +1,990 @@
+- **State:** Temporarily assigns AllocationResults to ResourceClaims during
+  the Assume phase.
+
+**PlacementBinPackingPlugin (New)** Implements `PlacementScorer`. Scores
```
I think this plugin can prevent the current PodGroup from fragmenting larger Levels, but it cannot prevent the current PodGroup from fragmenting smaller levels.   If the current podgroup uses fewer than all the nodes in this Placement, then there could be multiple `podsAssignment` options, and different options may have different fragmentation effects.  Since pod-at-a-time scheduling within the Placement is greedy,  we won't consider multiple `podsAssignment` options.

Its not clear to me that you can influence this enough using the per-pod `Score` plugins.

44past4 replied:
This is an important problem but in order to be able to define what does the lower/smaller levels fragmentation even mean we need to have at least two topology levels defined for a PodGroup - one lower/smaller which is a prefered/best-effort placement for a PodGroup and higher/larger which is a required placement for a PodGroup.

This is not in the scope of this KEP.

This being said when we will be adding support for multiple levels to address problem with lower/smaller levels fragmentation we will need to solve two subproblems - problem of scheduling pods within a higher/larger placement and problem of scoring of higher/larger placements.

When it comes to scheduling while generating potential placements we can keep track information about placements and their sub placements. When we go through all lower/smaller placements we can keep track of the number of pods which we were able to schedule in each placement. We can also extend the scoring function so that it would work with partial placements (placements which do not contain all pods within a PodGroup) so for each lower/smaller placements we can also get their scores together with the number of pods which we were able to fit in those. While checking the higher/larger placements instead of simply going pod by pod and checking all nodes within the placement for PodGroups which have only one PodSet we can:

- Check if there is any lower/smaller sub placement which can fit remaining pods from the PodGroup.
- If there are such lower/smaller sub placements we can select the one which can fit the fewest of pods and has the highest score and we can try scheduling there as many of the remaining pods as possible.
- If none of the lower/smaller sub placements can fit all remaining pods we can choose lower/smaller sub placements which can fit the highest number of pods and has the highest score and we can try to schedule there as many of the remaining pods as possible and repeat this process again.

This should lead to creating a pod assignment which uses as few of the sub placements as possible.

When it comes to scoring of those higher/larger placements we can extend the PodGroupAssignment struct to contain the information about the number of sub placements used by a given placement and their scores. This information could be used instead of normal bin packing logic to score such placements.

Apart from the API to define the multi-level placements all proposed interfaces should be able to support this logic but their implementation may need to change. All this should be considered in the future KEP for the multi-lever scheduling support.

erictune replied:
It is not necessary to have two levels on one PodGroup for this to be a problem.   It is only necessary that unrelated PodGroups can have different levels. It is also sufficient to have a PodGroup asking for 1 level, and some plain pods (e.g. from a Deployment).  

We will need some place to score a placement's effect on each Level that it touches.

erictune commented on keps/sig-scheduling/5732-topology-aware-workload-scheduling/README.md:
```diff
@@ -0,0 +1,990 @@
+// Placement represents a candidate domain for scheduling a PodSet.
+// It defines a set of nodes and/or proposed Dynamic Resource Allocation (DRA)
+// resource bindings necessary to satisfy the PodSet's requirements within that domain.
+type Placement struct {
```
What is the valid lifetime of a Placement object? Is it only one cycle of Workload Cycle?
If so, then we don't need to worry about updating Placements when we update our view of the cluster.
Maybe state this.

erictune commented on keps/sig-scheduling/5732-topology-aware-workload-scheduling/README.md:
```diff
@@ -0,0 +1,990 @@
+
+5. **Explicit Topology Definition:** Using a Custom Resource (NodeTopology) to
+   define and alias topology levels, removing the need for users to know exact
+   node label keys.
```
Explicit Topology Information also provided these things:
  - An explicit total order on levels within one Topology object (needed for  Multi-level and Prioritized Placement Scheduling)
  - An implicit label hierarchy requirement
      -  A level `n` label's nodes must be a subset of at only level `n+1` label.
      -  Useful for Multi-level placement, and for the hierarchical aggregated capacity optimization. 
   - A way to limit the number of levels
      - Limit by validating the list length in a Topology object.  
      - Limiting levels limits one term of algorithm complexity.  
  - A way to discourage creation of too many Topology objects
      - Only admins or cloud providers should create these usually.

Taken together, these properties make it easier to avoid the case where there are many more TAS-relevant labels (key/value pairs) than there are nodes.  

Also, while the initial algorithm is going to be greedy, in the sense that it examines one workload at a time, future algorithms may want to examine multiple workloads at once to find jointly optimal placements.  By allowing excess complexity in the structure of topology labels at the outset, we will limit our ability to do future global optimizations.

I think it is fine to leave Explicit Topology Definition out of Alpha.  However, before GA, we should either have beta Explicit Topology Definition, or have documented requirement for (1) the maximum number of label keys used for TAS, (2) partial order requirement over all TAS keys, and (3) nesting requirement for TAS labels.  

Otherwise, it will be hard to enforce those later.

johnbelamaric replied:
> we want to support DRA-based constraints eventually too, and these implicitly also imply certain topologies. They will not be defined by Topology definition anyway, so it will by definition cover only a subset of potential constraints. @johnbelamaric - for your thoughts too

If I understand what you mean, I would argue that DRA devices can actually *serve* as the way to define these explicit topologies.

wojtek-t replied:
> If I understand what you mean, I would argue that DRA devices can actually serve as the way to define these explicit topologies.

That's what I meant here.
The reason why I wrote "implicitly" is that even though it's explicit definition of a topology, it is hidden in DRA objects which is not where majority of users would expect the topology to defined.

And this is what I had on my mind writing the above comment - we already have a way to define the topology in some way via DRA - so I'm not 100% convinced yet that we need a separate dedicated topology definition.

erictune reviewed (commented):
Looks great overall!

44past4 committed:
Initial batch of fixes after reviews

44past4 commented on keps/sig-scheduling/5732-topology-aware-workload-scheduling/README.md:
```diff
@@ -0,0 +1,990 @@
+
+- **Selection:** Select the Placement with the highest score.
+
+- **Binding:** Proceed to bind pods to the assigned nodes and resources.
```
I have updated description with the information that pod-by-pod scheduling cycle with nominatedNodeNames set will be used here.

44past4 commented on keps/sig-scheduling/5732-topology-aware-workload-scheduling/README.md:
```diff
@@ -0,0 +1,990 @@
+    // GeneratePlacements generates a list of potential Placements for the given PodGroup and PodSet.
+    // Each Placement represents a candidate set of resources (e.g., nodes matching a selector)
+    // and potential DRA allocations where the PodSet might be scheduled.
+    GeneratePlacements(ctx context.Context, state *framework.CycleState, podGroup *PodGroupInfo, podSet *PodSetInfo, parentPlacements []*Placement) ([]*Placement, *framework.Status)
```
I have updated the PodGroupInfo so that it has a list of PodSetInfo. With this change it is sufficient to pass only PodGroupInfo to the GeneratePlacements function. When it comes to passing PodGroupInfo vs PodSetInfo the constraints defined on the PodGroup level apply to all pods within a PodGroup not just to a single PodSet so we want to have a single GeneratePlacements call per PodGroup and not a PodSet (I have updated the description of the function to make it clear).

44past4 committed:
Move DRA constraints support to beta

44past4 committed:
Fix TOC

wojtek-t reviewed (commented):
This is great now - I'm pretty aligned with the proposal now.

macsko commented on keps/sig-scheduling/5732-topology-aware-workload-scheduling/README.md:
```diff
@@ -0,0 +1,1015 @@
+
+- **Heterogeneous PodGroup Handling**: Sequential processing will be used
+  initially. Pods are processed sequentially; if any fail, the placement is
+  rejected.
```
What if we have more pods than gang's minCount? Shouldn't this phase just try to schedule as many pods from a pod group as possible (while fulfilling minCount constraint), even if they are not subsequent?

44past4 committed:
Smaller fixed based on review feedback.

44past4 committed:
Update Explicit Topology Definition description.

sanposhiho commented on keps/sig-scheduling/5732-topology-aware-workload-scheduling/README.md:
```diff
@@ -0,0 +1,1016 @@
+
+- **Action:** For each generated Placement:
+
+  1. Call `AssumePlacement` (binds context to the specific node selector/DRA
```
Please a bit more clarify. I don't understand what is actually happening here. What is "context" and what does "binds context to ..." mean and actual do? What is the difference from the existing `ReservePlugin`?

sanposhiho commented on keps/sig-scheduling/5732-topology-aware-workload-scheduling/README.md:
```diff
@@ -0,0 +1,1016 @@
+
+The algorithm proceeds in three main phases for a given Workload/PodGroup.
+
+#### Phase 1: Candidate Placement Generation
```
And, what if there's no valid placement? I suppose pods are moved back to the queue, but how are they going to be requeued? Should we require EnqueueExtensions for PlacementGenerator plugnis?

sanposhiho replied:
Also similarly, what if some placements are rejected by Phase 2? How should we requeue pods if they are back to the queue eventually?

In general though, if some nodes or topologies are rejected by somewhere in this topology aware scheduling, we definitely need EnqueueExtensions to be implemented.

44past4 replied:
I have updated PlacementGeneratorPlugin description with the information that it should implement EnqueueExtensions.

sanposhiho commented on keps/sig-scheduling/5732-topology-aware-workload-scheduling/README.md:
```diff
@@ -0,0 +1,1016 @@
+
+  2. Iterate through every pod in the PodGroup.
+
+  3. Run standard Pod-level Filter and Score.
```
Why do we need to run Score? It only checks the feasibility, right?

sanposhiho replied:
My above scenario is about the correctness of the scheduling. If the scheduler doesn't schedule pods but actually there's a space, then that is a bug. We can compromise the perf for example, but we can definitely not compromise 'correctness'.

wojtek-t replied:
> My above scenario is about the correctness of the scheduling. If the scheduler doesn't schedule pods but actually there's a space, then that is a bug. We can compromise the perf for example, but we can definitely not compromise 'correctness'.

I'm not sure I fully agree here. For me it's not really about correctness - the much closer analogy would be binpacking here. We try to "binpack" but didn't manage to find the solution.
I can definitely imagine a case where I need to schedule N pods on N nodes and there is exactly 1 correct placement out of N! potential ones (pod1 can be scheduled on nodes 1..N, pod 2 on nodes 1..N-1, pod 3 on 1..N-2, and so on). If we allow for arbitrary constraints across different pods, having 100% guarantee will be super hard and I'm not convinced we need it.
The question is where the boundary is though.

But I think we can continue this discussion without blocking Alpha of this KEP.

sanposhiho replied:
Yes, homogenous workloads should be easy.. 

> For me this is quite well aligned with the option 2 listed above.

Sorry, how is it aligned with the option2?
I mean, if we say we cannot make the scheduler work nicely enough with heterogeneous ones, it's like ...why we designed our API to be able to contain any amount of various pods in a single workload. Shouldn't we have designed it to be able to only build homogenous workloads in the first place? At the very least, we need to clearly define "we'll officially support this complexity/size of workloads in clusters at this scale" and then make a recommendation on the doc / build integration/scheduler_perf tests based on that. 

---

Also, we need to solve other problems with the scheduler making this incorrect "unschedulable" results:
- How we should requeue pods (workloads) in the scheduler queue: pods in unschedQ don't mean that the cluster doesn't have a place for them because a workload might be just unlucky. That means event based retries in general (QHints etc) would be completely invalid for workloads. OTOH, just keep retrying all such workload pods periodically sounds very inefficient.
- Cluster autoscaler: CA could create nodes unnecessarily for such unschedulable (but actually schedulable) workloads. That means, in the worst case, CA could keep creating nodes for the same workloads unnecessarily.

dom4ha replied:
I believe it should be in scope of KEP-4671 (see my comment above). Regarding QHints, I already argued that we should have time based wake up as a baseline solution and QHints as possible enhancement https://github.com/kubernetes/enhancements/pull/5730#discussion_r2697913908.

44past4 committed:
Added Plugin suffix to PlacementGenerator, PlacementState and PlacementScorer

44past4 commented on keps/sig-scheduling/5732-topology-aware-workload-scheduling/README.md:
```diff
@@ -0,0 +1,1016 @@
+    // GeneratePlacements generates a list of potential Placements for the given PodGroup.
+    // Each Placement represents a candidate set of resources (e.g., nodes matching a selector)
+    // and potential DRA allocations where the PodGroup might be scheduled.
+    GeneratePlacements(ctx context.Context, state *framework.CycleState, podGroup *PodGroupInfo, parentPlacements []*Placement) ([]*Placement, *framework.Status)
```
`FilterPlacements` would for me imply that the function gets some placements as input and returns some subsets of those placements as output whereas in this case even if we get some placements as `parentPlacements` we can returns more placements as output which are for instance a subdivision of the provided placements. Because of this the `GeneratePlacements` name seems more meaningful for me.

brejman commented on keps/sig-scheduling/5732-topology-aware-workload-scheduling/README.md:
```diff
@@ -0,0 +1,1016 @@
+    // GeneratePlacements generates a list of potential Placements for the given PodGroup.
+    // Each Placement represents a candidate set of resources (e.g., nodes matching a selector)
+    // and potential DRA allocations where the PodGroup might be scheduled.
+    GeneratePlacements(ctx context.Context, state *framework.CycleState, podGroup *PodGroupInfo, parentPlacements []*Placement) ([]*Placement, *framework.Status)
```
The plugin for TAS would need a list of nodes to discover the topology label values in order to generate placements.

In the proposal doc it was mentioned that PreFilter would be called before GeneratePlacements, giving plugins a chance to get the list of nodes, but I don't see a mention of it in the KEP.

Could you explain it in the KEP?

44past4 replied:
I have added information about GeneratePlacements being called after PreFilter scheduling phase to the Scheduling Algorithm Phases section of the KEP.

44past4 committed:
Updating README.md based on the comments

- Added requirement to PlacementGeneratorPlugin to implement EnqueueExtensions
- Added information about PlacementGeneratorPlugins to be called after PreFilter scheduling phase.
- Changed NodeAffinity to NodeSelector in Placement struc

dom4ha commented on keps/sig-scheduling/5732-topology-aware-workload-scheduling/README.md:
```diff
@@ -0,0 +1,1021 @@
+  1. Call `AssumePlacement` (binds context to the specific node selector/DRA
+     resources).
+
+  2. Iterate through every pod in the PodGroup.
```
This is the place where this algorithm leverages the workload scheduling cycle and scheduling algorithm defined there. So whatever algorithm we develop there, will be simply run under various option assumptions. So at the moment the algorithm indeed does iterate, but also use batching of similar pods. It currently does not try different scheduling orders, but whenever we consider it, it would be a part of workload scheduling cycle KEP.

So we should get:
- Call AssumePlacement (binds context to the specific node selector/DRA resources).
- Run default workload scheduling algorithm with the context
- Call RevertPlacement.

44past4 replied:
I have updated the description as suggested to indicate that the standard workload scheduling algorithm constrained to a given placement is used here.

dom4ha reviewed (commented):
LGTM

The algorithm is well described and straightforward. Obviously it will become more complicated once we consider more than a single domain and plugin, but for alpha any many use cases this is completely sufficient.

wojtek-t commented on keps/sig-scheduling/5732-topology-aware-workload-scheduling/README.md:
```diff
@@ -0,0 +1,1016 @@
+
+  2. Iterate through every pod in the PodGroup.
+
+  3. Run standard Pod-level Filter and Score.
```
I think that deciding how exactly it is feature gated is an implementation detail and doesn't belong to KEP.
That doesn't mean it's not important - it certainly is, but I'm assuming ensuring this is done reasonably is part of code approval.

44past4 committed:
Add prod readiness file

44past4 committed:
Production Readiness Review Questionnaire

44past4 committed:
Fixed spelling errors

wojtek-t reviewed (commented):
PRR looks fine to me - I will wait for approving for the SIG.

44past4 committed:
Update kep.yaml

44past4 committed:
Extend KEP with desiredCount.

44past4 committed:
Address comments from dom4ha

44past4 commented on keps/sig-scheduling/5732-topology-aware-workload-scheduling/README.md:
```diff
@@ -0,0 +1,1021 @@
+**TopologyPlacementPlugin (New)** Implements `PlacementGeneratorPlugin`. Generates
+Placements based on distinct values of the designated node label (TAS).
+
+**PlacementBinPackingPlugin (New)** Implements `PlacementScorerPlugin`. Scores
```
I have updated the KEP to actually have two scoring plugins for Placements. One which would check the number of pods which can be scheduled within a Placement which is required for desiredCount handling and the proposed PlacementBinPackingPlugin. PlacementBinPackingPlugin should work similar to the existing most allocated node resources scoring plugin preferring Placements with most requested resources.

44past4 committed:
Update README.md

44past4 committed:
Update README.md

44past4 committed:
Fixed Toc

dom4ha commented on keps/sig-scheduling/5732-topology-aware-workload-scheduling/README.md:
```diff
@@ -0,0 +1,966 @@
+    Name *string
+
+    // SchedulingConstraints defines group-level scheduling requirements,
+    // including topology and DRA colocation.
```
I think you can remove "and DRA collocation" part from this comment, as according to KEP https://github.com/kubernetes/enhancements/pull/5736, ResourceClaim won't be part of the SchedulingConstraints.

44past4 replied:
I have removed "and DRA collocation" as suggested.

dom4ha commented on keps/sig-scheduling/5732-topology-aware-workload-scheduling/README.md:
```diff
@@ -0,0 +1,966 @@
+Note: For the initial alpha scope, only a single TopologyConstraint will be
+supported.
+
+#### Basic Policy Extension
```
It was brought up in other comments in the original KEP https://github.com/kubernetes/enhancements/pull/5730#discussion_r2759161303, it makes sense to add DesiredCount for `Gang` policy as well for exactly the same reasons.

sanposhiho replied:
Hmm it's very confusing that we have `minCount` and `desiredCount` on `Gang`. Like, if I was a user looking at those fields for the first time, I'd definitely not get which is which.

dom4ha reviewed (commented):
LGTM

Added minor wording suggestions, but we need decision wrt we want to add DesiredCount to `Gang` policy as well.

44past4 committed:
Add desiredCount to Gang policy

44past4 committed:
Added cluster autoscaling support as requirement for beta

44past4 committed:
Fix phrasing

Co-authored-by: Dominik Marciński <gmidon@gmail.com>

44past4 committed:
Fix phrasing

Co-authored-by: Dominik Marciński <gmidon@gmail.com>

44past4 committed:
Updates from review.

44past4 committed:
Updates from review.

44past4 commented on keps/sig-scheduling/5732-topology-aware-workload-scheduling/README.md:
```diff
@@ -273,14 +273,22 @@ will be defined in a separate KEP:
+without imposing hard scheduling constraints.
+
 This feature will be gated behind a separate feature gate 
 (`WorkloadBasicPolicyDesiredCount`) to decouple it from the core Gang Scheduling
```
Fixed as suggested.

wojtek-t commented on keps/sig-scheduling/5732-topology-aware-workload-scheduling/README.md:
```diff
@@ -0,0 +1,966 @@
+type BasicSchedulingPolicy struct {
+	// DesiredCount is the expected number of pods that will belong to this
+	// PodGroup. This field is a hint to the scheduler to help it make better
+	// placement decisions for the group as a whole.
```
+1

wojtek-t commented on keps/sig-scheduling/5732-topology-aware-workload-scheduling/README.md:
```diff
@@ -0,0 +1,1021 @@
+
+    // PodGroupReplicaIndex is the index of the PodGroup replica, as defined in KEP-4671.
+    // This is relevant for PodGroups that have more than one replica.
+    PodGroupReplicaIndex int
```
OK - let's not block merging this KEP - once we make a decision, we should remember to update this KEP to reflect that.

wojtek-t commented:
/lgtm
/approve PRR
